addons/mrp/tests/test_replenish.py

```diff
@@ -72,6 +72,49 @@ def test_mrp_orderpoint_leadtime(self):
         # for manufacturing delay should be taken from the bom
         self.assertEqual("4.0 days", info.wh_replenishment_option_ids.lead_time)
 
+    def test_rr_picking_type_id(self):
+        """Check manufacturing order take bom according to picking type of the rule triggered by an
+        orderpoint."""
+
+        self.product_4.route_ids = self.env.ref('mrp.route_warehouse0_manufacture')
+        self.picking_type_2 = self.warehouse_1.manu_type_id.copy({'sequence': 100})
+        self.env['mrp.bom'].create({
+            'product_id': self.product_4.id,
+            'product_tmpl_id': self.product_4.product_tmpl_id.id,
+            'product_uom_id': self.uom_unit.id,
+            'product_qty': 4.0,
+            'type': 'normal',
+            'picking_type_id': self.picking_type_2.id,
+            'bom_line_ids': [
+                (0, 0, {'product_id': self.product_2.id, 'product_qty': 2}),
+                (0, 0, {'product_id': self.product_1.id, 'product_qty': 4})
+            ]})
+        self.rr = self.env['stock.warehouse.orderpoint'].create({
+            'name': 'Cake RR',
+            'location_id': self.warehouse_1.lot_stock_id.id,
+            'product_id': self.product_4.id,
+            'product_min_qty': 5,
+            'product_max_qty': 10,
+        })
+        self.assertTrue(self.product_4.bom_ids, 2)
+        self.assertTrue(self.product_4.bom_ids.picking_type_id, 2)
+
+        self.rr.action_replenish()
+        mo = self.env['mrp.production'].search([
+            ('product_id', '=', self.product_4.id),
+            ('picking_type_id', '=', self.warehouse_1.manu_type_id.id)
+        ])
+        self.assertTrue(mo)
+        mo.action_cancel()
+        self.warehouse_1.manufacture_pull_id.picking_type_id = self.picking_type_2
```

Comment:
Since we're changing it on the rule itself, it will pick the bom with the same picking type (as it did without your changes), and while it will apply the source/destination from the bom, they'll be the same as the one in the rule.

Maybe a better check would be to have only a single bom, have it match the rule picking type at first, then try again after changing the bom picking type and check if:
- The bom was selected
- The picking type, source and destination are the one from the bom